File: solutions/_2130_maximum_twin_sum_of_a_linked_list/Solution.java

package _2130_maximum_twin_sum_of_a_linked_list;

class Solution {
    public int pairSum(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        ListNode prev = null;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            ListNode tmp = slow.next;
            slow.next = prev;
            prev = slow;
            slow = tmp;
        }

        int res = 0;
        while (slow != null) {
            res = Math.max(res, slow.val + prev.val);
            prev = prev.next;
            slow = slow.next;
        }
        
        return res;
    }
}

File: solutions/_2130_maximum_twin_sum_of_a_linked_list/solution.js
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function (head) {
  let fast = head;
  let slow = head;
  let prev = null;

  while (fast && fast.next) {
    fast = fast.next.next;
    const tmp = slow.next;
    slow.next = prev;
    prev = slow;
    slow = tmp;
  }

  let res = 0;
  while (slow) {
    res = Math.max(res, prev.val + slow.val);
    prev = prev.next;
    slow = slow.next;
  }

  return res;
};

module.exports = { pairSum };

/*
https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/
*/
